extractor/filesystem/language/python/uvlock/uvlock.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/google/osv-scalibr/extractor/filesystem/osv"
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+)
+
+type uvLockPackageSource struct {
+	Virtual  string `toml:"virtual"`
+	Git      string `toml:"git"`
+	Registry string `toml:"registry"`
+}
+
+type uvLockPackage struct {
+	Name    string              `toml:"name"`
+	Version string              `toml:"version"`
+	Source  uvLockPackageSource `toml:"source"`
+
+	// uv stores "groups" as a table under "package" after all the packages, which due
+	// to how TOML works means it ends up being a property on the last package, even
+	// through in this context it's a global property rather than being per-package
+	Groups map[string][]uvOptionalDependency `toml:"optional-dependencies"`
+}
+
+type uvOptionalDependency struct {
+	Name string `toml:"name"`
+}
+type uvLockFile struct {
+	Version  int                               `toml:"version"`
+	Packages []uvLockPackage                   `toml:"package"`
+	Groups   map[string][]uvOptionalDependency `toml:"package.optional-dependencies"`
+}
+
+// Extractor extracts python packages from uv.lock files.
+type Extractor struct{}
+
+// Name of the extractor
+func (e Extractor) Name() string { return "python/uvlock" }
+
+// Version of the extractor
+func (e Extractor) Version() int { return 0 }
+
+// Requirements of the extractor
+func (e Extractor) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{}
+}
+
+// FileRequired returns true if the specified file matches uv lockfile patterns
+func (e Extractor) FileRequired(api filesystem.FileAPI) bool {
+	return filepath.Base(api.Path()) == "uv.lock"
+}
+
+// Extract extracts packages from uv.lock files passed through the scan input.
+func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	var parsedLockfile *uvLockFile
+
+	_, err := toml.NewDecoder(input.Reader).Decode(&parsedLockfile)
+
+	if err != nil {
+		return []*extractor.Inventory{}, fmt.Errorf("could not extract from %s: %w", input.Path, err)
+	}
+
+	packages := make([]*extractor.Inventory, 0, len(parsedLockfile.Packages))
+
+	var groups map[string][]uvOptionalDependency
+
+	// uv stores "groups" as a table under "package" after all the packages, which due
+	// to how TOML works means it ends up being a property on the last package, even
+	// through in this context it's a global property rather than being per-package
+	if len(parsedLockfile.Packages) > 0 {
+		groups = parsedLockfile.Packages[len(parsedLockfile.Packages)-1].Groups
+	}
+
+	for _, lockPackage := range parsedLockfile.Packages {
+		if lockPackage.Source.Virtual == "." {
+			continue
+		}
+
+		_, commit, _ := strings.Cut(lockPackage.Source.Git, "#")
+
+		pkgDetails := &extractor.Inventory{
+			Name:      lockPackage.Name,
+			Version:   lockPackage.Version,
+			Locations: []string{input.Path},
+			SourceCode: &extractor.SourceCodeIdentifier{
```

Comment:
Maybe we should just leave this struct nil if commit is ""

Reply:
I'll let you make the call on this since I'm not familiar with the rest of the codebase - my understanding is that by always setting this it would mean you don't need to check if its nil (since it's a pointer), which I personally like since the compiler doesn't do type narrowing (i.e. unlike say TypeScript, it won't tell you that you've forgotten to `nil` check your pointer).

However, that could very well be the expected value for this field in this situation 🤷

Reply:
In other extractors we currently only set SourceCode if there's some commit to write there: https://github.com/google/osv-scalibr/blob/main/extractor/filesystem/language/python/pdmlock/pdmlock.go#L87

In yet others we never set this struct so at the moment we don't gain anything from writing empty structs here. So for now I'd just avoid writing an empty struct and if we decide we should consistently write it we can make that change in a refactoring PR.